Reject pod when attachment limit is exceeded

torredil opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

When a CSI volume fails to attach due to a lack of attachment slots on the node, the pod can get stuck indefinitely in the `ContainerCreating` state - kubelet's volumeManager continuously retries to mount the volume, unaware that the underlying attachment error is terminal.

This change introduces a mechanism for the kubelet to detect this specific failure case. After the standard volume mount process times out, the volumeManager now investigates the cause of the failure by checking the status of the underlying VolumeAttachment object. I think this is the best place to perform the check as it resolves any high frequency polling issue.

If a terminal `ResourceExhausted` error is found, `WaitForAttachAndMount` returns a specific `ErrVolumeAttachLimitExceeded` error. The kubelet's SyncPod loop catches this error and cleanly terminates the pod by setting its phase to 'Failed'. This allows the pod's owning controller (such as a StatefulSet) to take corrective action.

--- 

I validated this CR by building kubelet with this patch and the external attacher with https://github.com/kubernetes-csi/external-attacher/pull/662 to populate the ErrorCode field when a CSI plugin returns the attach limit exceeded error.

After Kubelet's `verifyVolumesMountedFunc` timed out in `WaitForAttachAndMount`, it checked and saw that the volume failed to mount due to the limit exceeded error as reported in the VA object and rejected the pod, the SS then re-created the pod and scheduled it to another node.

```
kubectl get events                                                                                            
LAST SEEN   TYPE      REASON                          OBJECT                                                      MESSAGE
10m         Normal    Scheduled                       pod/app-0                                                   Successfully assigned default/app-0 to i-0f3f6a42ec7d87aa2
10m         Warning   FailedAttachVolume              pod/app-0                                                   AttachVolume.Attach failed for volume "pvc-6e49f588-a11a-4d3a-ad59-4e434361f644" : rpc error: code = ResourceExhausted desc = Attachment limit exceeded for volume
8m          Normal    Scheduled                       pod/app-0                                                   Successfully assigned default/app-0 to i-09c909676a26480cc
```

#### Which issue(s) this PR is related to:

[KEP-4876: Mutable CSINode Allocatable Property](https://github.com/torredil/enhancements/blob/master/keps/sig-storage/4876-mutable-csinode-allocatable/README.md)
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubelet now detects terminal CSI volume mount failures due to exceeded attachment limits on the node and marks the stateful pod as Failed, allowing its controller to recreate it. This prevents pods from getting stuck indefinitely in the `ContainerCreating` state.
```

torredil committed:
Reject pod when attachment limit is exceeded

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil commented:
cc: @jsafrane @gnufied @msau42 @SergeyKanzhelev

PR needs to be reviewed / approved by both sig-storage and sig-node.

gnufied commented on pkg/kubelet/kubelet.go:
```diff
@@ -2010,6 +2010,10 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 
 	// Wait for volumes to attach/mount
 	if err := kl.volumeManager.WaitForAttachAndMount(ctx, pod); err != nil {
+		if errors.Is(err, volumemanager.ErrVolumeAttachLimitExceeded) {
```
Can we also record the reason via `recordAdmissionRejection` ?

haircommander replied:
hmm this doesn't immediately seem like an admission rejection to me, this seems pretty far into the creation process to be considered admission...

haircommander replied:
I wouldn't call myself an expert here so I'd defer to others opinions but I would personally not record it as an admission rejection

gnufied replied:
hmm, that may be true. Is it safe to reject pods here then? The reason I asked here to record the metric is, because everytime we call `rejectPod`, we make sure to call `recordAdmissionRejection`. All `recordAdmissionRejection` does is - it records a metric.

There are already other resource exhausted errors which can cause rejection in kubelet. To me, volume limit exhaustion doesn't seem all that different.  We haven't started any of associated containers yet and hence it should be safe to fail the pod?

torredil replied:
Sergey recommended updating recordAdmissionRejection to include the extra metric label value, resolving.

torredil committed:
Record admission rejection

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil committed:
Fix pull-kubernetes-linter-hints

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil committed:
Fix AD Controller unit test failure

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil commented:
/label tide/merge-method-squash

gnufied commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -414,6 +417,29 @@ func (vm *volumeManager) WaitForAttachAndMount(ctx context.Context, pod *v1.Pod)
+			}
+		}
+	}
+
```
`getUnattachedVolumes` function should already return list of unattached volumes. It is just that, it returns a string whereas we need volumeSpec. So we can make that change. Something like - https://github.com/gnufied/kubernetes/commit/7427aa6bb841f4ad9092e9e4e6b3af4280d0a846

torredil replied:
Done in latest rev ✅

gnufied commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -414,6 +417,29 @@ func (vm *volumeManager) WaitForAttachAndMount(ctx context.Context, pod *v1.Pod)
+			}
+			if attachablePlugin.VerifyExhaustedResource(volumeToMount.VolumeSpec) {
+				// Return error to the kubelet, which will then trigger the pod termination logic.
+				return ErrVolumeAttachLimitExceeded
```
I think that it is a problem that, we aren't losing information which was available in the pod event before. So, rather than creating a hardcoded `ErrVolumeAttachLimitExceeded` value, can we not define a error type which still includes all the messages but type of error is `ErrVolumeAttachLimitExceeded`.

torredil replied:
Done in latest rev ✅

torredil committed:
Consolidate error handling logic in WaitForAttachAndMount

Signed-off-by: Eddie Torres <torredil@amazon.com>

gnufied commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -427,6 +436,28 @@ func (vm *volumeManager) WaitForAttachAndMount(ctx context.Context, pod *v1.Pod)
+				if attachablePlugin.VerifyExhaustedResource(volumeToMount.VolumeSpec) {
+					// Return error to the kubelet, which will then trigger the pod termination logic.
+					return &VolumeAttachLimitExceededError{
+						Err: err,
```
nit: can you compose err the way it is composed below? So something like:

```
fmt.Errorf(
			"unmounted volumes=%v, unattached volumes=%v, failed to process volumes=%v: %w",
			unmountedVolumes,
			unattachedVolumes,
			volumesNotInDSW,
			err)
```

My point is, user should still be able to see which of the volumes failed to attach and mount.

torredil replied:
Thank you! @gnufied, updated CR.

torredil committed:
Improve error context

Signed-off-by: Eddie Torres <torredil@amazon.com>

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -2010,6 +2010,13 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
+		if errors.As(err, &volumeAttachLimitErr) {
+			reasonStr := "VolumeAttachmentLimitExceeded"
+			kl.rejectPod(pod, reasonStr, "Node has insufficient volume attachment capacity")
+			recordAdmissionRejection(reasonStr)
```
I would suggest we have another reason for admission failure metric - please update the `recordAdmissionRejection` to include an extra metric label value

gnufied replied:
I think what @SergeyKanzhelev perhaps meant is, enum value in - https://github.com/AndrewSirenko/kubernetes/blob/vac-valid/pkg/kubelet/kubelet.go#L239 , so as metric gets emitted with proper reason, otherwise I think, it will simply show up as "Other" reason.

torredil replied:
@gnufied Yep, thats exactly what was implemented - see https://github.com/kubernetes/kubernetes/pull/132933/commits/e93a9edd23158114f3bcafdfa33c67c753aa5b71 (theres also a unit test for this in kubelet_test.go).

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -2010,6 +2010,13 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 
 	// Wait for volumes to attach/mount
 	if err := kl.volumeManager.WaitForAttachAndMount(ctx, pod); err != nil {
+		var volumeAttachLimitErr *volumemanager.VolumeAttachLimitExceededError
```
will there be other terminal failures from CSI? Would it be better to make it more generic?

torredil replied:
As Hemant mentioned, we currently don't have anything else and the existing name is more precise. Open to suggestions for making this more generic / future-proof, I was thinking something like `TerminalWaitForAttachAndMountError` if we want to make that change.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -2010,6 +2010,13 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
+		var volumeAttachLimitErr *volumemanager.VolumeAttachLimitExceededError
+		if errors.As(err, &volumeAttachLimitErr) {
+			reasonStr := "VolumeAttachmentLimitExceeded"
+			kl.rejectPod(pod, reasonStr, "Node has insufficient volume attachment capacity")
```
I am not sure if just setting the pod status to terminal state will make kubelet clean up all the resources that were allocated to the Pod. Did you test manually that clean up actually happened?

torredil replied:
Yes, this patch is manually tested - I validated that cleanup actually happened + the statefulset controller created a replacement pod once the existing one transitioned to a terminal, Failed state.

torredil replied:
Resolving thread - confirmed in latest revision that when `VolumeAttachLimitExceededError` is returned from `WaitForAttachAndMount`, the pod transitions to a terminal `Failed` state. Since `isTerminal` evaluates to true, this signals the kublet state machine to proceed to terminate / cleanup the pod.

Copied describe pod and kubelet logs below which show `SyncTerminatedPod` is executed, which is where cleanup for unmounting volumes, removing directories, and cleaning up cgroups before signaling that the pod's lifecycle is complete happens.

```
kubectl describe pod app 
Name:             app
Namespace:        default
Priority:         0
Service Account:  default
Node:             i-0f3f6a42ec7d87aa2/172.20.156.138
Start Time:       Fri, 18 Jul 2025 05:15:01 +0000
Labels:           <none>
Annotations:      kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container app
Status:           Failed
Reason:           VolumeAttachmentLimitExceeded
Message:          Pod was rejected: Node has reached its volume attachment limit, rejecting pod. unmounted volumes=[persistent-storage], unattached volumes=[persistent-storage], failed to process volumes=[]
```

```
3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1\") " pod="default/app"
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.165912   54821 pod_workers.go:1339] "Pod is terminal" pod="default/app" podUID="3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1" updateType="sync"
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.834422   54821 util.go:30] "No sandbox for pod can be found. Need to start a new one" pod="default/app"
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.838664   54821 kubelet.go:2159] "Pod termination stopped all running containers" pod="default/app" podUID="3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1"
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.847081   54821 reconciler_common.go:162] "operationExecutor.UnmountVolume started for volume \"kube-api-access-n9w8c\" (UniqueName: \"kubernetes.io/projected/3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1-kube-api-access-n9w8c\") pod \"3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1\" (UID: \"3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1\") "
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.852193   54821 operation_generator.go:781] UnmountVolume.TearDown succeeded for volume "kubernetes.io/projected/3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1-kube-api-access-n9w8c" (OuterVolumeSpecName: "kube-api-access-n9w8c") pod "3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1" (UID: "3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1"). InnerVolumeSpecName "kube-api-access-n9w8c". PluginName "kubernetes.io/projected", VolumeGIDValue ""
Jul 18 05:17:04 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:04.947768   54821 reconciler_common.go:299] "Volume detached for volume \"kube-api-access-n9w8c\" (UniqueName: \"kubernetes.io/projected/3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1-kube-api-access-n9w8c\") on node \"i-0f3f6a42ec7d87aa2\" DevicePath \"\""
Jul 18 05:17:05 i-0f3f6a42ec7d87aa2.us-west-2.compute.internal kubelet[54821]: I0718 05:17:05.838679   54821 kubelet.go:2275] "Pod is terminated and will need no more status updates" pod="default/app" podUID="3a6dc7e3-5fc0-43d3-8aa6-04054b4d65a1"
```

torredil committed:
Update admissionRejectionReasons to include VolumeAttachmentLimitExceededReason

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil committed:
Update status message

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil commented:
I've addressed all the comments and thoroughly validated latest rev. Besides adding more code coverage (which I plan to push on that front over the coming days), this is ready to be merged.

torredil committed:
Add TestWaitForAttachAndMountVolumeAttachLimitExceededError unit test

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil committed:
Add e2e test

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil committed:
Fix pull-kubernetes-linter-hints

Signed-off-by: Eddie Torres <torredil@amazon.com>

torredil commented:
Exercised new e2e test in [pull-kubernetes-e2e-kind-alpha-beta-features](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132933/pull-kubernetes-e2e-kind-alpha-beta-features/1947674018236600320): 
```
Kubernetes e2e suite: [It] [sig-storage] MutableCSINodeAllocatableCount [FeatureGate:MutableCSINodeAllocatableCount] [Alpha] [Feature:OffByDefault] Attach Limit Exceeded should transition pod to failed state when attachment limit exceeded
```

gnufied commented:
/lgtm

gnufied commented:
/assign @SergeyKanzhelev

SergeyKanzhelev commented on test/e2e/storage/csimock/mutable_csinode_allocatable.go:
```diff
@@ -50,67 +59,161 @@ var _ = utils.SIGDescribe("MutableCSINodeAllocatableCount", framework.WithFeatur
 
-		framework.ExpectNoError(err, "CSINode allocatable count was not updated to %d in time", updatedMaxVolumesPerNode)
-		framework.Logf("SUCCESS: MaxVolumesPerNode updated limit %d", updatedMaxVolumesPerNode)
+		f.It("should transition pod to failed state when attachment limit exceeded", func(ctx context.Context) {
```
It may be useful to add a test that runs this logic `500` times to ensure kubelet is not leaking any resources.

Also continuous and fast failures (without any back off) is a new expected behavior here. See  https://github.com/kubernetes/kubernetes/issues/108152

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -2010,6 +2011,12 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 
 	// Wait for volumes to attach/mount
 	if err := kl.volumeManager.WaitForAttachAndMount(ctx, pod); err != nil {
+		var volumeAttachLimitErr *volumemanager.VolumeAttachLimitExceededError
```
Continuing on this comment: https://github.com/kubernetes/kubernetes/pull/132933#discussion_r2211805558

It is not trivial to compare everything got prepared by that line and what and how being cleaned up later. The only worrisome thing I found is the DRA - it is prepared after this line, but unprepared all the time. So DRA plugin will see the Unprepare call for the resources that was never prepared.

This is similar to a pod entering crash loop backoff because of CSI failure and later being deleted. So DRA must be ready for this.

That said, from looking over all codepaths I can think of - this looks OK.

SergeyKanzhelev replied:
@pohly FYI

johnbelamaric replied:
I think this should be ok but we should make sure it's documented for driver authors.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
/hold

I am ok to add the stress-like test that creates 500 Pods as a follow up. Holding to allow to add it now, but feel free to unhold
